### DISCUSSION TITLE
Develop custom xplat controls and native platform dependencies

### DISCUSSION THREAD

**Discussion by workgroupengineering** [AvaloniaUI/Avalonia#12559] [General]
In https://github.com/AvaloniaUI/avalonia-dotnet-templates/issues/227 I request guidance on how to develop a custom xplat control using the native dependency.

As the number of controls available to avalonia this need becomes more and more important.

To abstract custom control from native platform dependencies, each developer uses the most convenient IoC for him. If you use controls  in your project with different orign/developer, you may have more IoCs dependencies.

The possible solutions that now come to my mind:

- Make AvaloniaLocator public again
- Create a surrogate for AvaloniaLocator which does not allow access to the private Avalonia APIs
- An attraction leayer for IoCs
- ...

cc @avaloniaui-team

**Comment by maxkatz6** [comment]
Please follow https://github.com/AvaloniaUI/Avalonia/discussions/11838 for some recommendations. There is no "single correct solution" though.

> To abstract custom control from native platform dependencies, each developer uses the most convenient IoC for him. 

IoC is absolutely not necessary for native controls. And there is a little point of adding IoC abstractions just for that.

**Reply by workgroupengineering** [reply]
Thank you, I'll take a look
